python/perf-kernels/flash-attention.py

```diff
@@ -851,23 +956,116 @@ def test_op_fwd(Z, H, N_CTX, D_HEAD, causal, use_bias, bias_type, qseqlen_not_eq
     if TORCH_HAS_FP8E5:
         q = q.to(torch_dtype)
         k = k.to(torch_dtype)
-    sm_scale = D_HEAD ** -0.5
-    dout = torch.randn_like(q, dtype=torch.float16)
+    o = torch.empty_like(q)
+
     # triton implementation
-    tri_out, _ = attention(q, k, v, causal, bias, sm_scale, 0, False, True) #dropout tested against SDPA only
+    tri_out, _ = attention(q, k, v, o, input_metadata)
     # reference implementation
     M = torch.tril(torch.ones((seqlen_q, seqlen_k), device="cuda"))
-    p = torch.matmul(q.half(), k.transpose(2, 3).half()) * sm_scale
+    scores = torch.einsum('bhqd,bhkd->bhqk', q, k).float() * sm_scale
     if causal:
-        p[:, :, M == 0] = float("-inf")
+        scores[:, :, M == 0] = float("-inf")
     if use_bias:
-        ref_bias, _ = prepare_bias(bias, Z, H, seqlen_q, seqlen_k)
-        p += ref_bias
-
-    p = torch.softmax(p.float(), dim=-1).half()
+        scores += input_metadata.bias
+    p = torch.softmax(scores.float(), dim=-1).half()
     ref_out = torch.matmul(p, v)
     # compare
-    torch.testing.assert_close(ref_out, tri_out, atol=4e-2, rtol=4e-2)
+    torch.testing.assert_close(ref_out, tri_out, atol=2e-2, rtol=2e-2)
+
+def varlen_input_helper(Z, HQ, HK, N_CTX, D_HEAD, dtype):
+    torch.manual_seed(20)
+    
+    # Random sequence lengths. Using N_CTX as kind of max of sum of individual seqs
+    max_seqlens = N_CTX // Z
+    seqlens_q = torch.randint(1, max_seqlens + 1, (Z,), dtype=torch.int32)
+    seqlens_k = torch.randint(1, max_seqlens + 1, (Z,), dtype=torch.int32)
+    max_seqlens_q = torch.max(seqlens_q).item()
+    max_seqlens_k = torch.max(seqlens_k).item()
+
+    # Calculate cumulative sequence lengths
+    cu_seqlens_q = torch.cat([torch.tensor([0], dtype=torch.int32), seqlens_q.cumsum(dim=0, dtype=torch.int32)])
+    cu_seqlens_k = torch.cat([torch.tensor([0], dtype=torch.int32), seqlens_k.cumsum(dim=0, dtype=torch.int32)])
+    cu_seqlens_q = cu_seqlens_q.to(device="cuda")
+    cu_seqlens_k = cu_seqlens_k.to(device="cuda")
+    # -1 because the last entry of cu_seqlens_q specifies the end of the last seq
+    num_ctxs = len(cu_seqlens_q) - 1
+
+    # Initialize q, k, v with variable lengths
+    total_q = cu_seqlens_q[-1].item()
+    total_k = cu_seqlens_k[-1].item()
+    q = torch.randn((total_q, HQ, D_HEAD), dtype=dtype, device="cuda").normal_(mean=0., std=0.5).requires_grad_()
+    k = torch.randn((total_k, HK, D_HEAD), dtype=dtype, device="cuda").normal_(mean=0., std=0.5).requires_grad_()
+    v = torch.randn((total_k, HK, D_HEAD), dtype=dtype, device="cuda").normal_(mean=0., std=0.5).requires_grad_()
+    sm_scale = D_HEAD**-0.5
+    input_metadata = MetaData(sm_scale=sm_scale)
+    input_metadata.set_varlen_params(cu_seqlens_q, cu_seqlens_k)
+    return q, k, v, input_metadata
+
+
+@pytest.mark.parametrize('Z, H, N_CTX, D_HEAD',
```

Comment:
Let's leave it as it is now but I'm expecting such patterns on UTs for full coverage,
```Python
# Copied from PyTorch
@parametrize("batch_size", [1, 8])
@parametrize("seq_len_q", [4, 8, 64, 143, 256, 512, 1024, 2048])
@parametrize("seq_len_k", [4, 8, 64, 128, 256, 587, 1024, 2048])
@parametrize("head_dim", [8, 16, 21, 32, 64, 72, 96, 128, 160, 192, 203, 256])
@parametrize("is_causal", [True, False])
@parametrize("dropout_p", [0.0, 0.22, 0.48])
@parametrize("dtype", [torch.float16, torch.bfloat16])
@parametrize("scale", [None, "l1"])
```

and for new UTs we can do it right for the first pass.

Reply:
Sounds good. We can make the tests more comprehensive once we split.